notebooks/01-tessellate-ipu-tile-api-basics.ipynb

```diff
@@ -0,0 +1,590 @@
+      "Platform=ipu\n",
+      "Number of devices=1\n",
+      "IpuDevice(id=0, num_tiles=8, version=ipu2)\n"
+     ]
+    }
+   ],
+   "source": [
+    "USE_IPU_MODEL = True\n",
+    "if USE_IPU_MODEL:\n",
+    "    print(\"Using IPU model\")\n",
+    "    from jax.config import config\n",
+    "\n",
+    "    config.FLAGS.jax_ipu_use_model = True\n",
+    "    config.FLAGS.jax_ipu_model_num_tiles = 8\n",
+    "\n",
+    "import jax\n",
+    "\n",
+    "# Check IPU hardware configuration\n",
+    "assert jax.default_backend() == \"ipu\"\n",
+    "print(f\"Platform={jax.default_backend()}\")\n",
+    "print(f\"Number of devices={jax.device_count()}\")\n",
+    "\n",
+    "devices = jax.devices()\n",
+    "print(\"\\n\".join([str(d) for d in devices]))\n",
+    "device0 = jax.devices()[0]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c7316e8b",
+   "metadata": {},
+   "source": [
+    "## Tile tensor sharding in TessellateIPU\n",
+    "\n",
+    "Prior to doing any compute on IPU tiles, you need to decide how to shard the data across tiles.  In normal usage, our IPU frameworks (PyTorch, TensorFlow, and JAX) will automatically decide on the optimal mapping of ML workloads using the Poplar compiler.\n",
+    "\n",
+    "TessellateIPU provides two primitives to allow the user to control this mapping directly:\n",
+    "* `tile_put_sharded`: Shards a tensor over the first axis between a set of tiles.\n",
+    "* `tile_put_replicated`: Replicates a tensor on a set of tiles.\n",
+    "\n",
+    "Both methods return a `TileShardedArray` Python object, which wraps a common JAX array and explicit (static) tile mapping.\n",
+    "A `TileShardedArray` tensor is always sharded over the first axis, and on-tile shards are contiguous in memory.\n",
+    "Reshaping or permuting the indices of such a tensor will use the all-to-all IPU exchange to efficiently rearrange the data across tiles. \n",
+    "\n",
+    "Here is an example showing how to use these two methods."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "0b69110c",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "First time compilation may take up to a minute... run the next cell to see the results\n"
+     ]
+    }
+   ],
+   "source": [
+    "import jax\n",
+    "import numpy as np\n",
+    "\n",
+    "from tessellate_ipu import tile_put_replicated, tile_put_sharded\n",
+    "\n",
+    "data = np.random.rand(3, 5).astype(np.float32)\n",
+    "\n",
+    "\n",
+    "@jax.jit\n",
+    "def compute_fn(data):\n",
+    "    # Shard data on tiles (0, 1, 3)\n",
+    "    t0 = tile_put_sharded(data, (0, 1, 3))\n",
+    "    # Replicate data on tiles (1, 3)\n",
+    "    t1 = tile_put_replicated(data, (1, 3))\n",
+    "    return t0, t1\n",
+    "\n",
+    "\n",
+    "print(\"First time compilation may take up to a minute... run the next cell to see the results\")\n",
```

Comment:
I cant see anywhere where you define where the poplar executable cache gets stored, is it stored in a default place? We use environment variables to save it in the right place. executable_cache_dir = os.getenv("POPLAR_EXECUTABLE_CACHE_DIR", "/tmp/exe_cache/")   Could you save the cache using this environment variable

Reply:
Poplar compilation compilation is actually not really an issue for this notebook (only a couple of seconds at most). The issue here is the compilation of TessellateIPU library: it relies on C++ custom ops, which are compiled on the fly at the first import of the library (and can't really be cached easily). I changed the sentence to make that a bit clearer, and I'll try to work on a solution in the future to speed that up!